include/app/action.hpp

```diff
@@ -60,11 +62,18 @@ class Action : public QObject {
     Q_OBJECT
 
    public:
-    Action(QString name, std::function<void()> func, QWidget *parent);
+    enum ActionState
+    {
+        Activated,
+        Deactivated,
+        Triggered
+    };
+    Action(QString name, std::function<void(ActionState)> func, QWidget *parent);
```

Comment:
seems a bit weird to muddy this up with the `ActionState` since it seems that's really only used for the AA interface calls through the ActionEventFilter?

Reply:
All actions set with shortcuts of single keys now are installed via the event filter - the action state allows for distinguishing between press and release, as well as compatibility with multi key shortcuts set with QShortcut (triggered state) - not sure how this could be cleaned up better.

Reply:
multikey shortcuts already work with dash no?

I guess at this point it's splitting the gpio triggers from the shortcut triggers? Because the gpio triggers wouldn't be handled in that even filter (or anything provided by qt)

Reply:
Yes, dash already handles multikey shortcuts, but with this PR, it handles single key shortcuts differently (event filter) so that we can detect the discrete press and release actions, rather than a single trigger point (the discrete states rn are used for android auto, but open the floodgates to implementations such as "key held" actions).

GPIO triggers remain activated via the old style as well.

Reply:
hmm not for this pr, but i think id like to see the gpio handling done through the even filter too then (idk if this is even possible or how it would look like tho)

Reply:
I'm not sure that's possible, unless QT sends events for gpio state changes (which I don't think it does)

Reply:
you can make custom qt events however so we can always roll our own (again just thinking out loud - feel free to disregard 🙂)